Main.hs

```diff
@@ -2,32 +2,122 @@ module Main where
 
 import Control.Monad (unless)
 import Text.Printf (printf)
+import Data.Either
 
-data Expr
+data Expr =
+  Num Double
+  | Sqrt Expr
+  | Add Expr Expr
+  | Sub Expr Expr
+  | Mul Expr Expr
+  | Div Expr Expr
+  | Pow Expr Expr
 
+x = Add (Num 4) (Mul (Num 4) (Num 6))
+y = Sqrt (Num 4)
+z = Add (Num 2) (Num 4)
 instance Show Expr where
-  show = undefined
+  show (Num a) = show a
+  show (Sqrt a) = show "sqrt(" ++ show a ++ ")" 
+  show (Add x y) = "(" ++ show x ++ " + " ++ show y ++ ")"
+  show (Sub x y) = "(" ++ show x ++ " - " ++ show y ++ ")"
+  show (Mul x y) = "(" ++ show x ++ " * " ++ show y ++ ")"
+  show (Div x y) = "(" ++ show x ++ " / " ++ show y ++ ")"
+  show (Pow x y) = "(" ++ show x ++ " ^ " ++ show y ++ ")"
 
 instance Eq Expr where
-  (==) = undefined
+  (==) x y = show x == show y
```

Comment:
It's not a good way to check for equality. You can do it without converting the expression to a string first.